au4/src/trackedit/internal/trackeditactionscontroller.cpp

```diff
@@ -237,6 +251,28 @@ void TrackeditActionsController::trackSplitAt(const ActionData& args)
     trackeditInteraction()->splitAt(trackId, playbackPosition);
 }
 
+void TrackeditActionsController::mergeSelectedOnTrack(const muse::actions::ActionData &args)
+{
+    IF_ASSERT_FAILED(args.count() == 3) {
+        return;
+    }
+
+    std::vector<TrackId> tracksIds = args.arg<std::vector<TrackId>>(0);
+    if (tracksIds.empty()) {
+        return;
+    }
+
+    secs_t begin = args.arg<secs_t>(1);
+    secs_t end = args.arg<secs_t>(2);
+    secs_t duration = end - begin;
+
+    for (const auto &trackId : tracksIds ) {
+        trackeditInteraction()->mergeSelectedOnTrack(trackId, begin, end);
+    }
+
+    pushProjectHistoryJoinState(begin, duration);
```

Comment:
Should be inside trackeditInteraction()->mergeSelectedOnTrack  
The reason is that the action controller is called only for processing of actions... 
But a service method can be called in other cases, in some scenario, from some API extensions, etc.
In this case, all the logic associated with this operation must be executed...
If we ever need to not modify the history in some case, then it is better to do this explicitly, in the service method add a flag "do not change to history"